Fixed issue with array and nullable type declarations in lambda parameters

acizmarik opened the pull request:
Resolves #1024. Additionally, I found out that it was not possible to use array types too. This PR should fix these issues and makes it possible to use constructs such as `int?`, `int[]`, `int?[][]` as lambda parameters.

exyi commented on src/DotVVM.Framework/Compilation/Parser/Binding/Parser/TypeDeclarationBindingParserNode.cs:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    [DebuggerDisplay("{DebuggerDisplay,nq}")]
+    public class TypeDeclarationBindingParserNode : BindingParserNode
```
This types seems overloaded - it should be abstract without the Type and Next properties as these are only used in certain subtypes.

Also, I'd call this a type reference - definition would be a `class X { }` syntax.

I'd imagine hierarchy likes this:

```csharp

abstract class TypeReferenceBindingParserNode
sealed class ArrayTypeReferenceBindingParserNode(TypeReferenceBindingParserNode elementType) : TypeReferenceBindingParserNode
sealed class NullableTypeReferenceBindingParserNode(TypeReferenceBindingParserNode innerType) : TypeReferenceBindingParserNode
sealed class ActualTypeReferenceBindingParserNode(BindingParserNode type) : TypeReferenceBindingParserNode
```

Including generic types into this hierarchy would be nice.

acizmarik commented:
I implemented the suggested class hierarchy. I also reworked how we treat generics so that they can be handled in a similar way - there is now a `GenericTypeReferenceBindingParserNode`.
